codec-redis/src/main/java/io/netty/handler/codec/redis/RedisEncoder.java

```diff
@@ -0,0 +1,154 @@
+    private static final byte[] CRLF = { '\r', '\n' };
+    private static final byte[] NULL_BULK_STRING = { '-', '1' };
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, RedisMessage msg, List<Object> out) throws Exception {
+        ByteBuf buf = ctx.alloc().buffer(calculateBufferLength(msg));
+        out.add(writeRedisMessage(buf, msg));
+    }
+
+    private static ByteBuf writeRedisMessage(ByteBuf buf, RedisMessage msg) {
+        if (msg instanceof StringRedisMessage) {
+            writeInlineByteBufMessage(buf, (StringRedisMessage) msg);
+        } else if (msg instanceof IntegerRedisMessage) {
+            writeIntegerMessage(buf, (IntegerRedisMessage) msg);
+        } else if (msg instanceof BulkStringRedisMessage) {
+            writeBulkStringMessage(buf, (BulkStringRedisMessage) msg);
+        } else if (msg instanceof ArrayHeaderRedisMessage) {
+            writeArrayHeaderObject(buf, (ArrayHeaderRedisMessage) msg);
+        } else if (msg instanceof ArrayRedisMessage) {
+            writeArrayMessage(buf, (ArrayRedisMessage) msg);
+        } else {
+            throw new CodecException("bad message: " + msg);
+        }
+        return buf;
+    }
+
+    private static void writeInlineByteBufMessage(ByteBuf buf, StringRedisMessage msg) {
+        buf.writeByte(msg.type().value());
+        buf.writeBytes(msg.content().getBytes(CharsetUtil.UTF_8));
+        buf.writeBytes(CRLF);
+    }
+
+    private static void writeIntegerMessage(ByteBuf buf, IntegerRedisMessage msg) {
+        buf.writeByte(msg.type().value());
+        buf.writeBytes(numberToBytes(msg.value()));
+        buf.writeBytes(CRLF);
+    }
+
+    private static void writeBulkStringMessage(ByteBuf buf, BulkStringRedisMessage msg) {
+        buf.writeByte(msg.type().value());
+        if (msg.isNull()) {
+            buf.writeBytes(NULL_BULK_STRING);
+        } else {
+            ByteBuf content = msg.content();
+            buf.writeBytes(numberToBytes(content.readableBytes()));
+            buf.writeBytes(CRLF);
+            buf.writeBytes(content);
+        }
+        buf.writeBytes(CRLF);
+    }
+
+    /**
+     * Write array header only without body. Use this if you want to write arrays as streaming.
+     */
+    private static void writeArrayHeaderObject(ByteBuf buf, ArrayHeaderRedisMessage msg) {
+        buf.writeByte(RedisMessageType.ARRAY.value());
+        if (msg.isNull()) {
+            buf.writeBytes(NULL_BULK_STRING);
+        } else {
+            buf.writeBytes(numberToBytes(msg.length()));
+        }
+        buf.writeBytes(CRLF);
+    }
+
+    /**
+     * Write full constructed array message.
+     */
+    private static void writeArrayMessage(ByteBuf buf, ArrayRedisMessage msg) {
+        buf.writeByte(msg.type().value());
+        if (msg.isNull()) {
+            buf.writeBytes(NULL_BULK_STRING);
+            buf.writeBytes(CRLF);
+        } else {
+            buf.writeBytes(numberToBytes(msg.children().size()));
```

Comment:
For integer/long to ASCII encoding, I don't know any simple method besides a greedy quotient encoding.